calyx-opt/src/passes/static_promotion.rs

```diff
@@ -428,73 +444,104 @@ impl Visitor for InferStaticTiming {
+            for assignment in s.group.borrow().assignments.iter() {
+                if !(assignment.dst.borrow().is_hole()
+                    && assignment.dst.borrow().name == "done")
+                {
+                    let static_s = Assignment::from(assignment);
+                    sg.borrow_mut().assignments.push(static_s);
                 }
-            })
-            .collect();
-
-        if let Some(time) = comp.control.borrow_mut().update_static(&comp_lat) {
-            let mut go_ports = comp
-                .signature
-                .borrow()
-                .find_all_with_attr(ir::NumAttr::Go)
-                .collect_vec();
+            }
+            let s_enable = ir::StaticControl::Enable(ir::StaticEnable {
+                group: Rc::clone(&sg),
+                attributes: s.attributes.clone(),
+            });
+            return Ok(Action::change(ir::Control::Static(s_enable)));
+        }
+        Ok(Action::Continue)
+    }
 
-            // Add the latency information for the component if the control program
-            // is completely static and there is exactly one go port.
-            if go_ports.len() == 1 {
-                let go_port = go_ports.pop().unwrap();
-                let mb_time =
-                    go_port.borrow().attributes.get(ir::NumAttr::Static);
-
-                if let Some(go_time) = mb_time {
-                    if go_time != time {
-                        let msg1 = format!("Annotated latency: {}", go_time);
-                        let msg2 = format!("Inferred latency: {}", time);
-                        let msg = format!(
-                        "Invalid \"static\" latency annotation for component {}.\n{}\n{}",
-                        comp.name,
-                        msg1,
-                        msg2
-                    );
-                        return Err(Error::malformed_structure(msg)
-                            .with_pos(&go_port.borrow().attributes));
-                    }
-                } else {
-                    go_port
-                        .borrow_mut()
-                        .attributes
-                        .insert(ir::NumAttr::Static, time);
+    fn invoke(
+        &mut self,
+        s: &mut ir::Invoke,
+        _comp: &mut ir::Component,
+        _sigs: &LibrarySignatures,
+        comps: &[ir::Component],
+    ) -> VisResult {
+        if let ir::CellType::Component { name } = s.comp.borrow().prototype {
+            for c in comps {
+                if c.name == name && c.is_static() {
+                    let s_inv = ir::StaticInvoke {
+                        comp: Rc::clone(&s.comp),
+                        inputs: s.inputs.clone(),
+                        outputs: s.outputs.clone(),
+                        latency: c.latency.unwrap().get(),
+                        attributes: s.attributes.clone(),
+                        ref_cells: s.ref_cells.clone(),
```

Comment:
Too many `clone`s here. Aren't we going to delete the `ir::Invoke` node? Can we detach the various fields without cloning them?